subprojects/packagefiles/libtomcrypt/0003.patch

```diff
@@ -0,0 +1,141 @@
+diff --git a/src/hashes/sha3.c b/src/hashes/sha3.c
+index c6faa0b..567d45e 100644
+--- a/src/hashes/sha3.c
++++ b/src/hashes/sha3.c
+@@ -174,6 +174,102 @@ int sha3_shake_init(hash_state *md, int num)
+    if (num != 128 && num != 256) return CRYPT_INVALID_ARG;
+    XMEMSET(&md->sha3, 0, sizeof(md->sha3));
+    md->sha3.capacity_words = (unsigned short)(2 * num / (8 * sizeof(ulong64)));
++   md->sha3.suffix = 0x1F;
++   return CRYPT_OK;
++}
++
++static unsigned long cshake_left_encode(unsigned long num, unsigned char *buf)
++{
++    unsigned int n, i;
++    size_t v;
++
++    for (v = num, n = 0; v && (n < sizeof(unsigned long)); n++, v >>= 8) {
++        /* empty */
++    }
++    if (n == 0)
++        n = 1;
++
++    for (i = 1; i <= n; i++)
++    {
++        buf[i] = (unsigned char)(num >> (8 * (n - i)));
++    }
++    buf[0] = (unsigned char)n;
++
++    return n + 1;
++}
++
++int sha3_cshake_init(hash_state *md, int num,
++        const unsigned char *func, unsigned long funclen,
++        const unsigned char *custom, unsigned long customlen)
++{
++   unsigned char buf[168];
++   unsigned long len, total = 0;
++   unsigned long rate;
++   LTC_ARGCHK(md != NULL);
++   if (num != 128 && num != 256) return CRYPT_INVALID_ARG;
++   rate = (num == 128) ? 168 : 136;
++   XMEMSET(&md->sha3, 0, sizeof(md->sha3));
++   md->sha3.capacity_words = (unsigned short)(2 * num / (8 * sizeof(ulong64)));
++
++   if (funclen == 0 && customlen == 0) {
++      /* no function, no customization => regular SHAKE */
++      md->sha3.suffix = 0x1F;
++   } else {
++      md->sha3.suffix = 0x04;
++
++      /* start bytepad */
++      len = cshake_left_encode(rate, buf);
++      sha3_process(md, buf, len);
++      total += len;
++
++      /* process function name */
```

Comment:
Would it make sense to define a function `sha3_shake_process_string` (or similarly named) to process both the left encoded length of the string and then the string itself? This is used in several places.

Reply:
Sure, I extracted this into a `cshake_left_encode_and_process_str` function.